src/main/java/org/gwtproject/tools/cldr/LocalesNativeNamesProcessor.java

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright 2010 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.gwtproject.tools.cldr;
+
+import com.squareup.javapoet.*;
+import org.gwtproject.i18n.shared.cldr.*;
+import org.gwtproject.i18n.shared.GwtLocale;
+
+import javax.lang.model.element.Modifier;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.text.Collator;
+import java.util.*;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.isNull;
+import static org.gwtproject.tools.cldr.DateTimeFormatInfoProcessor.FORMATS;
+
+/**
+ * Extract list formatting information from CLDR data.
+ */
+public class LocalesNativeNamesProcessor extends Processor {
+
+    public static final Logger LOGGER = Logger.getLogger(LocalesNativeNamesProcessor.class.getName());
+
+    private final RegionLanguageData regionLanguageData;
+
+    /**
+     * Set of canonical language codes which are RTL.
+     * this is a copy/paste from gwt
+     */
+    private static final Set<String> RTL_LOCALES = new HashSet<>();
+
+    static {
+        // TODO(jat): get this from CLDR data.
+        RTL_LOCALES.add("ar");
+        RTL_LOCALES.add("fa");
+        RTL_LOCALES.add("he");
+        RTL_LOCALES.add("ps");
+        RTL_LOCALES.add("ur");
+    }
+
+    private static final Map<String, String> OVERRIDE_LOCALES_NATIVE_NAMES = new HashMap<>();
+
+    static {
+        OVERRIDE_LOCALES_NATIVE_NAMES.put("ssy", "Saho");
+    }
+
+    public LocalesNativeNamesProcessor(File outputDir, InputFactory cldrFactory, LocaleData localeData) {
+        super(outputDir, cldrFactory, localeData);
+        regionLanguageData = new RegionLanguageData(cldrFactory);
+    }
+
+    @Override
+    protected void cleanupData() {
+        localeData.copyLocaleData("en", "default", "territory", "languages", "scripts", "variants");
+        // Generate a sort order before removing duplicates
+        for (GwtLocale locale : localeData.getNonEmptyLocales("territory")) {
+            // TODO(jat): deal with language population data that has a script
+            Map<String, String> map = localeData.getEntries("territory", locale);
+            List<String> countryCodes = new ArrayList<String>();
+            for (String regionCode : map.keySet()) {
+                // only include real country codes
+                if (!"ZZ".equals(regionCode) && regionCode.length() == 2) {
```

Comment:
i saw another place where this was mentioned - maybe move this into a constant, and the constant can have the comment about what ZZ vs XX mean?